galley/pkg/fs/fssource_test.go

```diff
@@ -0,0 +1,348 @@
+// limitations under the License.
+
+package fs
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"istio.io/istio/galley/pkg/runtime/resource"
+)
+
+var mixerYAML = `
+apiVersion: "config.istio.io/v1alpha2"
+kind: denier
+metadata:
+  name: some.mixer.denier
+spec:
+  status:
+    code: 7
+    message: Not allowed
+---
+apiVersion: "config.istio.io/v1alpha2"
+kind: checknothing
+metadata:
+  name: some.mixer.checknothing
+spec:
+---
+apiVersion: "config.istio.io/v1alpha2"
+kind: rule
+metadata:
+  name: some.mixer.rule
+spec:
+  match: destination.labels["app"] == "someapp"
+  actions:
+  - handler: some.denier
+    instances: [ some.checknothing ]
+`
+
+var mixerPartYAML = `
+apiVersion: "config.istio.io/v1alpha2"
+kind: denier
+metadata:
+  name: some.mixer.denier
+spec:
+  status:
+    code: 7
+    message: Not allowed
+---
+apiVersion: "config.istio.io/v1alpha2"
+kind: checknothing
+metadata:
+  name: some.mixer.checknothing
+spec:
+`
+
+var virtualServiceYAML = `
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: route-for-myapp
+spec:
+  hosts:
+  - some.example.com
+  gateways:
+  - some-ingress
+  http:
+  - route:
+    - destination:
+        host: some.example.internal
+`
+
+var virtualServiceChangedYAML = `
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: route-for-myapp-changed
+spec:
+  hosts:
+  - some.example.com
+  gateways:
+  - some-ingress
+  http:
+  - route:
+    - destination:
+        host: some.example.internal
+`
+var fst *fsTestSourceState
+
+type fsTestSourceState struct {
+	ConfigFiles map[string][]byte
+	rootPath    string
+}
+
+func (fst *fsTestSourceState) testSetup(t *testing.T) {
+	var err error
+
+	fst.rootPath, err = ioutil.TempDir("", "configPath")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for name, content := range fst.ConfigFiles {
+		err = ioutil.WriteFile(filepath.Join(fst.rootPath, name), content, 0600)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func (fst *fsTestSourceState) writeFile() error {
+	for name, content := range fst.ConfigFiles {
+		err := ioutil.WriteFile(filepath.Join(fst.rootPath, name), content, 0600)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (fst *fsTestSourceState) deleteFile() error {
+	for name := range fst.ConfigFiles {
+		err := os.Remove(filepath.Join(fst.rootPath, name))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (fst *fsTestSourceState) testTeardown(t *testing.T) {
+	err := os.RemoveAll(fst.rootPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestNewSource(t *testing.T) {
+	fst = &fsTestSourceState{
+		ConfigFiles: map[string][]byte{},
+	}
+	fst.testSetup(t)
+	defer fst.testTeardown(t)
+	s, err := New(fst.rootPath)
+	if err != nil {
+		t.Fatalf("Unexpected error found: %v", err)
+	}
+
+	if s == nil {
+		t.Fatal("expected non-nil source")
+	}
+}
+func TestFsSource_InitialScan(t *testing.T) {
```

Comment:
Nit : You can make these tests fully declarative, which allows us to add more tests quickly as needed.